server/utils/dnsCacheManager.ts

```diff
@@ -0,0 +1,261 @@
+        family: 4,
+        timestamp: Date.now(),
+        ttl: 0,
+      };
+    }
+
+    const cached = this.cache.get(hostname);
+    if (cached) {
+      const age = Date.now() - cached.timestamp;
+      const ttlRemaining = Math.max(0, cached.ttl - age);
+
+      if (ttlRemaining > 0) {
+        this.stats.hits++;
+        logger.debug(`DNS cache hit for ${hostname}`, {
+          label: 'DNSCache',
+          address: cached.address,
+          family: cached.family,
+          age,
+          ttlRemaining,
+        });
+        return cached;
+      }
+
+      // soft expiration using stale entry while refreshing
+      if (age < this.hardTtlMs) {
+        this.stats.hits++;
+        logger.debug(`Using stale DNS cache for ${hostname}`, {
+          label: 'DNSCache',
+          address: cached.address,
+          family: cached.family,
+          age,
+          ttlRemaining,
+        });
+
+        // revalidation
+        this.resolveWithTtl(hostname)
+          .then((result) => {
+            this.cache.set(hostname, {
+              address: result.address,
+              family: result.family,
+              timestamp: Date.now(),
+              ttl: result.ttl,
+            });
+            logger.debug(`DNS cache refreshed for ${hostname}`, {
+              label: 'DNSCache',
+              address: result.address,
+              family: result.family,
+              ttl: result.ttl,
+            });
+          })
+          .catch((error) => {
+            logger.error(
+              `Failed to refresh DNS for ${hostname}: ${error.message}`
+            );
+          });
+
+        return cached;
+      }
+
+      // hard expiration: remove stale entry
+      this.cache.delete(hostname);
+    }
+
+    this.stats.misses++;
+    try {
+      const result = await this.resolveWithTtl(hostname);
+
+      const dnsCache: DnsCache = {
+        address: result.address,
+        family: result.family,
+        timestamp: Date.now(),
+        ttl: result.ttl,
+      };
+
+      this.cache.set(hostname, dnsCache, { ttl: this.hardTtlMs });
+      logger.debug(`DNS cache miss for ${hostname}, cached new result`, {
+        label: 'DNSCache',
+        address: dnsCache.address,
+        family: dnsCache.family,
+        ttl: result.ttl,
+      });
+
+      return dnsCache;
+    } catch (error) {
+      throw new Error(`DNS lookup failed for ${hostname}: ${error.message}`);
+    }
+  }
+
+  private async resolveWithTtl(
+    hostname: string
+  ): Promise<{ address: string; family: number; ttl: number }> {
+    if (
+      !this.resolver ||
+      typeof this.resolver.resolve4 !== 'function' ||
+      typeof this.resolver.resolve6 !== 'function'
+    ) {
+      throw new Error('Resolver is not initialized');
+    }
+
+    try {
+      const [ipv4Records, ipv6Records] = await Promise.allSettled([
+        this.resolver.resolve4(hostname, { ttl: true }),
+        this.resolver.resolve6(hostname, { ttl: true }),
+      ]);
```

Comment:
I'm not sure this is a good idea to wait for both resolution.
You should probably follow the same algorithm than [Happy Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs):
- start ipv6 resolution
- wait a very short delay
- start ipv4 resolution in parallel
- returns the entry of whichever resolves first

Reply:
The RFC recommends 150-250ms: https://datatracker.ietf.org/doc/html/rfc6555#section-5.5